vars/buildDockerAndPublishImage.groovy

```diff
@@ -13,96 +13,73 @@ def call(String imageName, Map config=[:]) {
+
+  podTemplate(
+    inheritFrom: 'jnlp-linux',
+    yaml: yamlPodDef,
+    defaultContainer: 'builder'
+  ) {
+    withEnv([
+      "BUILD_DATE=${dockerConfig.buildDate}",
+      "IMAGE_NAME=${dockerConfig.imageName}",
+      "DOCKERFILE=${dockerConfig.dockerfile}",
+      "PLATFORM=${dockerConfig.platform}",
+    ]) {
+      // Automatic tagging on principal branch is not enabled by default
+      if (semVerEnabled) {
+        stage('Next Version') {
           container('next-version') {
-            script {
-              nextVersion = sh(script:'jx-release-version', returnStdout: true).trim()
-              if (dockerConfig.metadataFromSh != '') {
-                metadata = sh(script: "${dockerConfig.metadataFromSh}", returnStdout: true).trim()
-                nextVersion = nextVersion + metadata
-              }
-            }
-          }
+            nextVersion = sh(script:'jx-release-version', returnStdout: true).trim()
+            if (dockerConfig.metadataFromSh != '') {
+              metadata = sh(script: "${dockerConfig.metadataFromSh}", returnStdout: true).trim()
+              nextVersion = nextVersion + metadata
+            } // if
+          } // container
           echo "Next Release Version = $nextVersion"
-        } // steps
-      } // stage
+        } // stage
+      } // if
+
 
       stage('Prepare') {
-        environment {
-          DOCKER_REGISTRY = credentials("${dockerConfig.credentials}")
-        }
-        steps {
+        withCredentials([usernamePassword(credentialsId: dockerConfig.credentials, passwordVariable: 'DOCKER_REGISTRY_PSW', usernameVariable: 'DOCKER_REGISTRY_USR')]) {
+          // Logging in on the Dockerhub helps to avoid request limit from DockerHub
           sh 'echo "${DOCKER_REGISTRY_PSW}" | img login -u "${DOCKER_REGISTRY_USR}" --password-stdin'
-          script {
-            // Retrieve Makefile
-            writeFile file: 'Makefile', text: makefileContent
-          }
-        } // steps
+
+          // The makefile to use must come from the pipeline to avoid a nasty user trying to exfiltrate data from the build
+          // Even though we have mitigation trhough the multibranch job config only allowed to build PRs from repo. contributors
+          writeFile file: 'Makefile', text: makefileContent
+        } // withCredentials
       } // stage
 
       stage("Lint") {
-        steps {
-          sh 'make lint'
-        } // steps
-        post {
-          always {
-            recordIssues enabledForFailure: true, tool: hadoLint(pattern: 'hadolint.json')
-          }
-        } // post
-      } //stage
+        int lintStatus = sh(returnStatus: true, script: 'make lint')
+        recordIssues enabledForFailure: true, tool: hadoLint(pattern: 'hadolint.json')
+        if (lintStatus > 0) {
+          currentBuild.result = 'FAILURE'
+          error('Lint Failed')
+        }
```

Comment:
```suggestion
        sh 'make lint'
        recordIssues enabledForFailure: true, tool: hadoLint(pattern: 'hadolint.json')
```
(If `make lint` failed, no reason to also record a Jenkins publisher.)

Reply:
you may want to record it so that it gets reported to github with the direct issues rather than having to look through the build

the idiom in scripted would normally be
```suggestion
        try {
          sh 'make lint'
        } finally {
          recordIssues enabledForFailure: true, tool: hadoLint(pattern: 'hadolint.json')
        }
```

Reply:
You _can_ do that, if the `make lint` output is illegible. Just make sure `recordIssues` does not produce an error if no `hadolint.json` file exists.

Reply:
👍

Reply:
(for the record, I was mistakenly using the `createShMock` method which led me to these overcomplicated scripts. With your pointers and some tries/retries I understand where my error is, and your feedback are clear!)